src/mi/backtrace.c

```diff
@@ -69,12 +70,52 @@ unw_backtrace (void **buffer, int size)
   if (unlikely (tdep_trace (&cursor, buffer, &n) < 0))
     {
       unw_getcontext (&uc);
-      return slow_backtrace (buffer, size, &uc);
+      return slow_backtrace (buffer, size, &uc, 0);
     }
 
   return n;
 }
 
+int
+unw_backtrace2 (void **buffer, int size, unw_context_t* uc2, int flag)
+{
+  if (size == 0)
+    return 0;
+    
+  if (uc2 == NULL)
+    return unw_backtrace(buffer, size);
+
+  unw_cursor_t cursor;
+  // need to copy, because the context will be modified by tdep_trace
+  unw_context_t uc = *(unw_context_t*)uc2;
+
+  if (unlikely (unw_init_local2 (&cursor, &uc, flag) < 0))
+    return 0;
+
+  // get the first ip from the context
+  unw_word_t ip;
+
+  if (unw_get_reg (&cursor, UNW_REG_IP, &ip) < 0)
+    return 0;
+
+  buffer[0] =  (void *) (uintptr_t)ip;
+
+  // update buffer info to collect the rest of the IPs
+  buffer = buffer+1;
+  int remaining_size = size-1;
+
+  int n = remaining_size;
+
+  // returns the number of frames collected by tdep_trace or slow_backtrace
+  // and add 1 to it (the one we retrieved above)
+  if (unlikely (tdep_trace (&cursor, buffer, &n) < 0))
+    {
+      return slow_backtrace (buffer, remaining_size, &uc, flag) + 1;
```

Comment:
@gleocadie its hard for me to test this, but if `tdep_trace` modifies `uc` (as commented above) and then fails, it seems to me it would no longer be safe to pass `&uc` to slow_backtrace. Do you agree?

Reply:
`tdep_trace` modifies `uc` but not the `unw_step` (at least I did not see any hint of modification).
The `slow_backtrace` relies only on `unw_step`.
From my tests (only on x86_64 arch) I had no crash.

Reply:
is it ok to use the modified `&uc` in slow_backtrace's `if (unlikely (unw_init_local2 (&cursor, uc, flag) < 0))` though?

Reply:
oh I see your point. maybe a another copy should be done.